### PR TITLE
home-manager: verify username and home directory

### LINKED ISSUE #4019: bug: Running switch with sudo puts symbolic links into `/root` eventhough `homeDirectory` is defined
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

I was running `home-manager switch` as sudo, mistakenly

```sh
sudo home-manager switch --flake '.#'
```

and it turns out the command updates symlinked files defined with `home.file` at `/root`.
Eventhough my `home.nix` contains:

```nix
  home = {
    inherit username;
    homeDirectory = "/home/${username}";
  };
```

Is it expected? I definitely didn't.

### Maintainer CC

_No response_

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 5.15.94, NixOS, 22.11 (Raccoon), 22.11.20230219.0cf4274`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - nixpkgs: `/nix/store/hbg5gjcr6bzya30advl28k4ynafhclhn-source`
```

### PR DESCRIPTION
**Pull request by rycee**
### Description

The generation activation script should be run by the user specified in `home.username` and `home.homeDirectory`. If some other user runs the activation script, then files may end up in the wrong place or with the wrong owner.

This commits adds a check early in the activation script that verifies that the running user match the user in the configuration.

Fixes #4019

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

### CODE REVIEW

File: modules/lib-bash/activation-init.sh [hunk 1]
```diff
@@ -88,6 +88,24 @@ function setupVars() {
...
+  if [[ $USER != $expectedUser ]]; then
+    _iError 'Error: USER is set to "%s" but we expect "%s"' "$expectedUser" "$USER"
+    exit 1
+  fi
+}
+
+function checkHomeDirectory() {
+  local expectedHome="$1"
+
+  if [[ $HOME != $expectedHome ]]; then
```
Comment (ncfavier) [inline modules/lib-bash/activation-init.sh:103]:
```suggestion
  if [[ $HOME -ef $expectedHome ]]; then
```

File: modules/lib-bash/activation-init.sh [hunk 2]
```diff
@@ -88,6 +88,24 @@ function setupVars() {
     fi
 }
 
+function checkUsername() {
+  local expectedUser="$1"
+
+  if [[ $USER != $expectedUser ]]; then
```
Comment (ncfavier) [inline modules/lib-bash/activation-init.sh:94]:
```suggestion
  if [[ $USER != "$expectedUser" ]]; then
```

RHS of `!=` is subject to globbing

File: modules/lib-bash/activation-init.sh (hunk 1 above)
Comment (rycee) [reply modules/lib-bash/activation-init.sh]: Thanks 🙂

File: modules/lib-bash/activation-init.sh (hunk 2 above)
Comment (rycee) [reply modules/lib-bash/activation-init.sh]: Thanks 🙂

**Pull request merged**